compiler/src/dmd/backend/machobj.d

```diff
@@ -2957,3 +2962,170 @@ int dwarf_eh_frame_fixup(int dfseg, targ_size_t offset, Symbol *s, targ_size_t v
+
+    private
+    {
+        /**
+         * This is the absolute minimum supported version of macOS (64 bit) for DMD,
+         * as documented at: https://dlang.org/dmd-osx.html#requirements
+         * NOTE: Versions earlier than 10.7 do not support thread local storage.
+         */
+        enum fallbackOSVersion = Version(10, 9).encode;
+
+        /// The first minor version that uses the `build_version_command`.
+        enum firstMinorUsingBuildVersionCommand = 14;
+
+        /// `true` if the `build_version_command` load command should be used.
+        bool useBuild;
+
+        /// The `build_version_command` load command.
+        build_version_command buildVersionCommand;
+
+        /// The `version_min_command` load command.
+        version_min_command versionMinCommand;
+    }
+
+    /**
+     * Initializes the VersionCommand.
+     *
+     * Params:
+     *  os = the version of the operating system
+     */
+    this(Version os)
+    {
+        useBuild = os.minor >= firstMinorUsingBuildVersionCommand;
+
+        const encodedOs = os.isValid ? os.encode : fallbackOSVersion;
+
+        const build_version_command buildVersionCommand = { minos: encodedOs };
+        const version_min_command versionMinCommand = { version_: encodedOs };
+
+        this.buildVersionCommand = buildVersionCommand;
+        this.versionMinCommand = versionMinCommand;
+    }
+
+    /// Returns: the size of the load command.
+    size_t size()
+    {
+        return useBuild ? build_version_command.sizeof : version_min_command.sizeof;
+    }
+
+    /// Returns: the data for the load command.
+    const(void)* data() return
+    {
+        return useBuild ? cast(const(void)*) &buildVersionCommand : cast(const(void)*) &versionMinCommand;
+    }
+}
+
+/// Holds an operating system version or a SDK version.
+immutable struct Version
+{
+    ///
+    int major;
+
+    ///
+    int minor;
+
+    ///
+    int build;
+
+    /// Returns: `true` if the version is valid
+    bool isValid() pure nothrow @nogc @safe
+    {
+        return major >= 10 && major < 100 &&
+            minor >= 0 && minor < 100 &&
+            build >= 0 && build < 100;
+    }
+}
+
+/**
+ * Returns the given version encoded as a single integer.
+ *
+ * Params:
+ *  version_ = the version to encode. Needs to be a valid version
+ *      (`version_.isValid`)
+ *
+ * Returns: the encoded version
+ */
+int encode(Version version_) pure @nogc @safe
+in
+{
+    assert(version_.isValid);
+}
+do
+{
+    with (version_)
+        return major * 2^^16 + minor * 2^^8 + build * 2^^0;
+}
+
+unittest
+{
+    assert(Version(10, 14, 0).encode == 658944);
+    assert(Version(10, 14, 1).encode == 658945);
+    assert(Version(10, 14, 6).encode == 658950);
+    assert(Version(10, 14, 99).encode == 659043);
+
+    assert(Version(10, 15, 6).encode == 659206);
+
+    assert(Version(10, 16, 0).encode == 659456);
+    assert(Version(10, 16, 6).encode == 659462);
+
+    assert(Version(10, 17, 0).encode == 659712);
```

Comment:
dscanner is complaining about lack of underscores in long integer literal constants: `dmd/backend/machobj.d(3079:41)[warn]: Use underscores to improve number constant readability.`

Reply:
The only thing that would improve readability would be to make it a hex or binary literal instead. As this is a unittest, making this pretty is mostly meaningless.

Reply:
Oh, I totally agree. I was just pointing out why the build kite style check failed. Will that continue to make the style check fail if this is merged though?

Reply:
Lets see if dscanner accepts hex then...